### PR TITLE
modify pytype_test to run from within pytype too, and support python3

### PR DESCRIPTION
**Pull request by martindemello**
pytype needs to run in a python2.7 venv, but also to have a python3.6 interpreter available to shell out to. this PR adds support for python3 stdlib testing to pytype_tests, and installs a python3.6 interpreter in the travis config.

### CODE REVIEW

**Comment by matthiaskramm** [comment]
Restarted mypy test, now that https://github.com/python/typeshed/pull/1818 is in.

File: tests/pytype_test.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env python
-"""Test runner for typeshed.
+r"""Test runner for typeshed.
```
Comment (matthiaskramm) [inline tests/pytype_test.py]: "r"? Like, in a regexp?

File: tests/pytype_test.py [hunk 2]
```diff
@@ -15,17 +15,33 @@
...
+                    help='Path to directory with pytype and pytd executables.')
+# Set to true to print a stack trace every time an exception is thrown.
+parser.add_argument('--print-stderr', type=bool, default=False,
+                    help='Print stderr every time an error is encountered.')
+# We need to invoke python3.6. The default here works with our travis tests.
+parser.add_argument('--python36-exe', type=str,
+                    default='/opt/python/3.6/bin/python3.6',
+                    help='Path to a python 3.6 interpreter.')
+
+Dirs = collections.namedtuple('Dirs', ['pytype', 'typeshed'])
```
Comment (matthiaskramm) [inline tests/pytype_test.py]: s/Dirs/Config/ ?

File: tests/pytype_test.py [hunk 3]
```diff
@@ -83,34 +106,56 @@ def communicate(self):
         return self.results
 
 
+def _get_relative(filename):
+    top = filename.find('stdlib')
```
Comment (matthiaskramm) [inline tests/pytype_test.py]: adding '/' after 'stdlib' might make this a tiny bit more robust.

File: tests/pytype_test.py [hunk 4]
```diff
@@ -83,34 +106,56 @@ def communicate(self):
...
+    try:
+        BinaryRun([pytd_exe, '-h']).communicate()
+    except OSError:
+        # See if pytd is named pytd_tool instead.
+        pytd_exe = os.path.join(dirs.pytype, 'pytd_tool')
+        try:
+            BinaryRun([pytd_exe, '-h']).communicate()
+        except OSError:
+            print('Cannot run pytd. Did you install pytype?')
+            return 0, 0
```
Comment (matthiaskramm) [inline tests/pytype_test.py]: I think this might be a bit more readable if you did a "can_run" method that does the try/BinaryRun/except part and returns True/False, since then you can do:

```
if can_run(config.pytype, 'pytd', '-h'):
  pytd_exe = os.path.join(config.pytype, 'pytd')
elif can_run(config.pytype, 'pytd_tool', '-h'):
  pytd_exe = os.path.join(config.pytype, 'pytd_tool')
else:
  print ('Cannot run pytd. Did you install pytype?')
  return 0, 0
```

File: tests/pytype_test.py [hunk 5]
```diff
@@ -119,16 +164,21 @@ def pytype_test(args):
     while 1:
         while files and len(running_tests) < args.num_parallel:
             f = files.pop()
+            run_cmd = [
+                pytype_exe,
+                '--typeshed-location=%s' % dirs.typeshed,
+                '--module-name=%s' % _get_module_name(f),
+                '--convert-to-pickle=%s' % os.devnull
+            ]
```
Comment (matthiaskramm) [inline tests/pytype_test.py]: Move this below to where it's used?

File: tests/pytype_test.py [hunk 6]
```diff
@@ -137,15 +187,20 @@ def pytype_test(args):
...
         max_code = max(max_code, code)
         runs += 1
 
         if code:
-            print(stderr)
+            if args.print_stderr:
+                print(stderr)
             errors += 1
+            bad.append((_get_relative(test_run.args[-1]),
+                        stderr.rstrip().rsplit('\n', 1)[-1]))
```
Comment (matthiaskramm) [inline tests/pytype_test.py]: This discards all lines but the last? Please add a comment.

File: tests/pytype_test.py (hunk 1 above)
Comment (martindemello) [reply tests/pytype_test.py]: r for raw string, since it has a literal backslash within it

File: tests/pytype_test.py (hunk 2 above)
Comment (martindemello) [reply tests/pytype_test.py]: i'll rename it if i add any more configuration options; right now it's just directories (saves having to name the fields config.pytype_dir and config.typeshed_dir)

File: tests/pytype_test.py (hunk 3 above)
Comment (martindemello) [reply tests/pytype_test.py]: done

File: tests/pytype_test.py (hunk 4 above)
Comment (martindemello) [reply tests/pytype_test.py]: done

File: tests/pytype_test.py (hunk 6 above)
Comment (martindemello) [reply tests/pytype_test.py]: done

File: tests/pytype_test.py (hunk 5 above)
Comment (martindemello) [reply tests/pytype_test.py]: it's used just below; the if block appends to it.

File: tests/pytype_test.py [hunk 7]
```diff
@@ -119,16 +168,21 @@ def pytype_test(args):
...
+                pytype_exe,
+                '--typeshed-location=%s' % dirs.typeshed,
+                '--module-name=%s' % _get_module_name(f),
+                '--convert-to-pickle=%s' % os.devnull
+            ]
+            if 'stdlib/3' in f:
+                run_cmd += [
+                    '-V 3.6',
+                    '--python_exe=%s' % args.python36_exe
+                ]
```
Comment (matthiaskramm) [inline tests/pytype_test.py]: move the definition of run_cmd and the "if" block below, to where it's used? Right now, it's confusing that you're building up this command, but then you're not using it in the "f not in pytype_run" case.

File: tests/pytype_test.py (hunk 5 above)
Comment (matthiaskramm) [reply tests/pytype_test.py]: see below

**Comment by matthiaskramm** [comment]
@JelleZijlstra: Could you take a look whether you're OK with the changes to .travis.yml?

File: .travis.yml [hunk 8]
```diff
@@ -11,12 +11,13 @@ matrix:
...
     - python: "2.7"
       env: TEST_CMD="./tests/pytype_test.py --num-parallel=4"
+      sudo: true
 
 install:
   # pytype needs py-2.7, mypy needs py-3.3+. Additional logic in runtests.py
   - if [[ $TRAVIS_PYTHON_VERSION == '3.6-dev' ]]; then pip install -U flake8==3.3.0 flake8-bugbear>=17.3.0 flake8-pyi>=17.1.0; fi
   - if [[ $TRAVIS_PYTHON_VERSION == '3.5' ]]; then pip install -U git+git://github.com/python/mypy git+git://github.com/python/typed_ast; fi
-  - if [[ $TRAVIS_PYTHON_VERSION == '2.7' ]]; then pip install -U git+git://github.com/google/pytype; fi
+  - if [[ $TRAVIS_PYTHON_VERSION == '2.7' ]]; then pip install -U git+git://github.com/google/pytype; wget https://s3.amazonaws.com/travis-python-archives/binaries/ubuntu/14.04/x86_64/python-3.6.tar.bz2; sudo tar xjf python-3.6.tar.bz2 --directory /; fi
```
Comment (JelleZijlstra) [inline .travis.yml]: I don't really like this use of wget. Could we instead run in a 3.6 environment, but then use the system python2.7 to run pytype?

File: tests/pytype_blacklist.txt [hunk 9]
```diff
@@ -79,3 +79,13 @@ stdlib/3/tokenize.pyi # parse only
 stdlib/3/types.pyi # parse only
 stdlib/3/urllib/error.pyi # parse only
 stdlib/3/urllib/request.pyi # parse only
+stdlib/3/collections/abc.pyi # parse only
```
Comment (JelleZijlstra) [inline tests/pytype_blacklist.txt]: Just curious, why were these added?

File: .travis.yml (hunk 8 above)
Comment (martindemello) [reply .travis.yml]: we need to be running in a 2.7 venv otherwise pytype's C extensions try to build against the wrong python header files and fail. i tried a bunch of things, and wgetting the binaries and extracting them as root was the only thing that worked. (this is not an entirely new thing; travis runs the same wget command when we ask for python 3.5, which is how i found out which server to get the binaries from)

File: tests/pytype_blacklist.txt (hunk 9 above)
Comment (martindemello) [reply tests/pytype_blacklist.txt]: they're temporarily broken while i add some features to pytype; i just didn't want them to block this PR. i'll do a separate PR in a few days cleaning out a lot of the blacklist.

File: tests/pytype_test.py (hunk 7 above)
Comment (martindemello) [reply tests/pytype_test.py]: done

File: tests/pytype_test.py [hunk 10]
```diff
@@ -87,34 +109,60 @@ def communicate(self):
...
+        pytd_exe = os.path.join(dirs.pytype, 'pytd_tool')
+    else:
         print('Cannot run pytd. Did you install pytype?')
         return 0, 0
 
-    skip, parse_only = load_blacklist()
+    skip, parse_only = load_blacklist(dirs)
     wanted = re.compile(r'stdlib/.*\.pyi$')
     skipped = re.compile('(%s)$' % '|'.join(skip))
     parse_only = re.compile('(%s)$' % '|'.join(parse_only))
```
Comment (matthiaskramm) [inline tests/pytype_test.py]: It's not directly related to this PR, but while you're here: Could you handle the case of `skip` or `parse_only` being empty lists?

**Comment by matthiaskramm** [comment]
@JelleZijlstra: Let me know if you have any other ideas/comments about the `wget` issue. Otherwise, I plan on merging this version on Monday.

**Comment by JelleZijlstra** [comment]
I am fine with this version.

File: tests/pytype_test.py (hunk 10 above)
Comment (martindemello) [reply tests/pytype_test.py]: done
